crates/kitsune_p2p/types/src/tx2/tx2_promote.rs

```diff
@@ -0,0 +1,701 @@
+#![allow(clippy::new_ret_no_self)]
+//! Promote a tx2 transport backend to a tx2 transport frontend.
+
+use crate::tx2::tx2_backend::*;
+use crate::tx2::tx2_frontend::tx2_frontend_traits::*;
+use crate::tx2::tx2_frontend::*;
+use crate::tx2::tx2_utils::*;
+use crate::tx2::*;
+use crate::*;
+use futures::future::{BoxFuture, FutureExt};
+use futures::stream::{Stream, StreamExt};
+use std::collections::HashSet;
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+/// Promote a tx2 transport backend to a tx2 transport frontend.
+pub fn tx2_promote(backend: BackendFactory, max_cons: usize) -> EpFactory {
+    EpFactory(Arc::new(PromoteFactory { max_cons, backend }))
+}
+
+// -- private -- //
+
+/// Limit the number of active channels allowed per connection.
+const CHANNELS_PER_CONNECTION: usize = 3;
```

Comment:
This feels a little strange but maybe I'm misunderstanding. Does this map to quic channels? If so in holochain we have lots of places where we open channels to send a message / wait for a response. I think there could be potential deadlocks if we are trying to open a connection but we already have 3 open waiting for responses? 
Maybe 3 is too low in that case?

Reply:
2 things:

1. We're switching to uni-streams, so we won't wait on the responses here, we'll just send the next req or res and move on
2. We're streaming frames over the uni-streams, so they can send multiple messages before closing, so at some point we can do some batching if we find it to be more efficient.